peps/pep-9999.rst

```diff
@@ -0,0 +1,546 @@
+not specified.
+
+Akin to ``Final``, read-only attributes do not influence the mutability of
+the assigned object. Immutable ABCs and containers may be used in combination with
+``ReadOnly`` to prevent mutation of such values.
+
+Syntax
+------
+
+``ReadOnly`` can be used at class-level or within ``__init__`` to mark individual
+attributes read-only:
+
+.. code-block:: python
+
+    class Book:
+        id: ReadOnly[int]
+
+        def __init__(self, id: int, name: str) -> None:
+            self.id = id
+            self.name: ReadOnly = name
+
+The explicit type in ``ReadOnly[<type>]`` can be omitted if the declaration has
+an initializing value. A type checker should apply its usual type inference
+rules to determine the type of ``name``.
+
+If an attribute is already implied to be read-only, like in the frozen :ref:`classes`,
+explicit declarations should be permitted and seen as equivalent, except ``Final``
+additionally forbids overriding in subclasses:
+
+.. code-block:: python
+
+    @dataclass(frozen=True)
+    class Point:
+        x: ReadOnly[int]
+        y: Final[int]
+
+Initialization
+--------------
+
+Assignment to a ``ReadOnly`` attribute can only occur as a part of the declaration,
+or within ``__init__`` of the same class. There is no restriction to how many
+times the attribute can be assigned to.
+
+.. code-block:: python
+
+    from collections import abc
+    from typing import ReadOnly
+
+
+    class Band:
+        name: str
+        songs: ReadOnly[list[str]]
+
+        def __init__(self, name: str, songs: abc.Iterable[str] | None = None) -> None:
+            self.name = name
+            self.songs = []
+
+            if songs is not None:
+                # multiple assignments during initialization are fine
+                self.songs = list(songs)
+
+        def clear(self) -> None:
+            # error: assignment to read-only "songs" outside initialization
+            self.songs = []
+
+
+    band = Band(name="Bôa", songs=["Duvet"])
+    band.name = "Python"  # ok: "name" is not read-only
+    band.songs = []  # error: "songs" is read-only
+    band.songs.append("Twilight")  # ok: list is mutable
+
+
+    class SubBand(Band):
+        def __init__(self) -> None:
+            # error: cannot assign to a read-only attribute of base class
+            self.songs = []
+
+An initializing value at a class level can serve as a `flyweight <https://en.wikipedia.org/wiki/Flyweight_pattern>`_
+default for instances:
+
+.. code-block:: python
+
+    class Patient:
+        number: ReadOnly = 0
+
+        def __init__(self, number: int | None = None) -> None:
+            if number is not None:
+                self.number = number
+
+This feature should also be supported by ``Final`` attributes. Specifically,
+``Final`` attributes initialized in a class body **should no longer** imply ``ClassVar``,
+and should remain assignable to within ``__init__``.
```

Comment:
What is the motivation for this change? I'm opposed to changing the semantics of final, since this seems to have some problematic side effects. Currently class-level final attributes can be used to define constants, where the literal value is known statically:
```py
class Foo:
    FOO: Final = 5
    BAR: Final = 6
    
    def __init__(...): ...

x: Literal[5] = Foo.FOO  # OK
```
(Mypyc even treats `Foo.FOO` as a compile-time constant at runtime, replacing references to `Foo.FOO` with 5.)

If we make the change as suggested in the PEP draft, it seems to imply that one of these is true:
1. To infer the type of a class-level final attribute, a type checker will need to combine types from assignments in two scopes: the class body and the `__init__` method.
2. The code above would start generating a type error (since the inferred type of `Foo.FOO` would likely have to be `int` if we must allow an arbitrary `__init__`), breaking backward compatibility.

I think (2) above is not reasonable due to the backward compatibility break. 

I also argue that (1) would also be quite a drastic change, since it adds "spooky action at a distance" where two assignments in two different parts of a file impact the type of an attribute. Here's what I'm thinking about:
```py
class Foo
    foo: Final = 5  # No longer can use initializer to infer type

    # assume lots of other stuff
    # ...
    
    def __init__(self, foo: int | None) -> None:
        if foo is not None:
            self.foo = foo  # This assignment means that "foo" has type "int"
```
I don't think there is a precedent for the above kind of type inference behavior, and it would be quite complex to implement in mypy at least. Also, to preserve backward compatibility, it would require new kind of type inference logic at least in mypy, again because there is no precedent for this kind of type inference behavior. This seems unnecessarily complicated considering that `ReadOnly` seems to be a better fit for the above use case.

In summary, I think that this change seems out of place in the PEP, and increases implementation complexity significantly without a clear benefit that I can see.

Reply:
I didn't initially plan on touching `Final`. I might have got the wrong idea, but @erictraut's initial [feedback](https://discuss.python.org/t/expanding-readonly-to-normal-classes-protocols/67359/45) seemed to suggest changing the spec of `Final` to allow assignment at class level and within `__init__` [^1]:
> I’m not convinced that we should allow double initialization — in both the class body and in __init__. Is there a good reason to allow both initialization techniques within the same class? Interestingly, mypy allows this for Final but pyright does not. The typing spec is currently silent on whether this should be allowed. Regardless of where we land on this decision, it would be good for this PEP to provide clarity for both ReadOnly and Final.

...hence the change. Likely should've discussed this more thoroughly.

[^1]: Under the precondition that `ReadOnly` obtains the feature

Reply:
Eric's feedback definitely did suggest clarifying the spec for `Final` in this PEP, though it didn't advocate for any particular conclusion or change.

My feeling is that we don't need to change the behavior of `Final` to make it the same as `ReadOnly`; that causes churn, and I think it's OK if they behave differently. I see this difference as deriving naturally from their different meaning; `Final` is really supposed to mean "this is the only assignment to this name, it is final", whereas `ReadOnly` just means "once an instance of this class is constructed, this attribute can't be assigned to."

I (weakly) feel also that it's a bit out of scope for this PEP to clarify existing underspecified aspects of `Final`, when those could just as well be clarified in a separate PR to the typing spec, approved by the typing council. But I'm not strongly opposed to this. If we do it, I would suggest collecting everything in this PEP related to `Final` in a single section, which presents a comparison of the semantics of `Final` and `ReadOnly`, and any clarifications we want to make to the specification of `Final`, all together.